sys/uio.c

```diff
@@ -69,16 +74,21 @@ ssize_t writev(int fd, const struct iovec *iov, int iovcnt)
 	unsigned int i;
 	ssize_t n, tot_len = 0;
 
-	if (-1 == checkv(iov, iovcnt))
-		return -EINVAL;
+	if (-1 == checkv(iov, iovcnt)) {
+		errno = EINVAL;
+		return -1;
+	}
 
 	/* TODO: file locking */
 
 	for (i = 0; i < iovcnt; ++i) {
 		if (iov[i].iov_len > 0) {
 			n = write(fd, iov[i].iov_base, iov[i].iov_len);
-			if (n == -1)
+			if (n < 0) {
+				if ((tot_len > 0 && errno == EINTR) || errno == EAGAIN)
```

Comment:
As above, shouldn't we return `-1` in case of `EAGAIN`?

Reply:
Yes if tot_len is zero. If tot_len is non-zero we should return number of bytes written and return EAGAIN on next call. For context check out this code from dropbear:

	written = writev(fd, iov, io_count);

	if (written < 0) {
		if (errno != EINTR && errno != EAGAIN) {
			TRACE(("channel IO write error fd %d %s", fd, strerror(errno)))
			close_chan_fd(channel, fd, SHUT_WR);
			return DROPBEAR_FAILURE;
		}
	} else {
		int cbuf_written = MIN(circ_len1+circ_len2, (unsigned int)written);
		cbuf_incrread(cbuf, cbuf_written);
		if (morelen) {
			*morelen = written - cbuf_written;
		}
		channel->recvdonelen += written;
	}

Reply:
Ok, so your condition should look like that:
`if (tot_len > 0 && (errno == EINTR || errno == EAGAIN))`
i.e. you should return `tot_len` only, if it's not zero.